source/stdx/collections/rcstring.d

```diff
@@ -0,0 +1,244 @@
+///
+module stdx.collections.rcstring;
+
+import stdx.collections.common;
+import stdx.collections.array;
+
+debug(CollectionRCString) import std.stdio;
+
+version(unittest)
+{
+    import std.experimental.allocator.mallocator;
+    import std.experimental.allocator.building_blocks.stats_collector;
+    import std.experimental.allocator : RCIAllocator, RCISharedAllocator,
+           allocatorObject, sharedAllocatorObject;
+    import std.algorithm.mutation : move;
+
+    private alias SCAlloc = StatsCollector!(Mallocator, Options.bytesUsed);
+}
+
+///
+struct RCString
+{
+private:
+    Array!ubyte _support;
+    AllocatorHandler _allocator;
+
+    /// Constructs the ouroboros allocator from allocator if the ouroboros
+    // allocator wasn't previously set
+    /*@nogc*/ nothrow pure @safe
+    bool setAllocator(A)(ref A allocator)
+    if (is(A == RCIAllocator) || is(A == RCISharedAllocator))
+    {
+        if (_allocator.isNull)
+        {
+            auto a = typeof(_allocator)(allocator);
+            move(a, _allocator);
+            return true;
+        }
+        return false;
+    }
+
+public:
+
+    /**
+     * Constructs a qualified rcstring that will use the provided
+     * allocator object. For `immutable` objects, a `RCISharedAllocator` must
+     * be supplied.
+     *
+     * Params:
+     *      allocator = a $(REF RCIAllocator, std,experimental,allocator) or
+     *                  $(REF RCISharedAllocator, std,experimental,allocator)
+     *                  allocator object
+     *
+     * Complexity: $(BIGOH 1)
+     */
+    this(A, this Q)(A allocator)
+    if (!is(Q == shared)
+        && (is(A == RCISharedAllocator) || !is(Q == immutable))
+        && (is(A == RCIAllocator) || is(A == RCISharedAllocator)))
+    {
+        debug(CollectionRCString)
+        {
+            writefln("RCString.ctor: begin");
+            scope(exit) writefln("RCString.ctor: end");
+        }
```

Comment:
I wonder whether it wouldn't be better to create a DebugProxy that automatically adds this code to all forwarded functions.
It would reduce the code size in the allocators quite considerably.